### PR TITLE
Correct Observable.filteri bug and optimise Observable.mapi

### PR DESCRIPTION
**Pull request by anton-pt**
Corrected the definition of Observable.filteri which previously only took a predicate of type 'T -> bool and simplified the definition of Observable.mapi, also improving performance.

### CODE REVIEW

File: src/Observable.fs [hunk 1]
```diff
@@ -560,7 +560,7 @@ module Observable =
     /// Filters the observable elements of a sequence based on a predicate by 
     /// incorporating the element's index
     let filteri predicate (source: IObservable<'T>)  = 
-        Observable.Where( source, Func<_,_> predicate )
+        Observable.Where( source, Func<_,_,_> (fun i x -> predicate x i) )
```
Comment (forki) [inline src/Observable.fs]: I think we should have a test or something for this...

Comment (panesofglass) [inline src/Observable.fs]: Yes, and great catch!

File: src/Observable.fs [hunk 2]
```diff
@@ -1028,12 +1028,7 @@ module Observable =
     /// Maps the given observable with the given function and the 
     /// index of the element
     let mapi (f:int -> 'Source -> 'Result) (source:IObservable<'Source>) =
-        source
-        |> Observable.scan ( fun (i,_) x -> (i+1,Some(x))) (-1,None)
-        |> Observable.map
-            (   function
-                | i, Some(x) -> f i x
-                | _, None    -> invalidOp "Invalid state"   )
+        Observable.Select (source, Func<_,_,_> (fun i x -> f x i))
```
Comment (panesofglass) [inline src/Observable.fs]: Wow; I wish I had seen that earlier. Thank you!

**Comment by anton-pt** [comment]
Hmmm, so I was thinking of bringing in FsCheck to do the tests but to do so I need to upgrade the project to .NET 4.5. Any objections?

**Comment by cloudRoutine** [comment]
by all means please do

**Comment by panesofglass** [comment]
:+1:

**Comment by anton-pt** [comment]
Anything else needed before this is ready to be merged in? Sorry: I don't have a lot of experience contributing to OSS and doing unit testing....

File: src/FSharp.Control.Reactive.fsproj [hunk 3]
```diff
@@ -9,7 +9,7 @@
     <OutputType>Library</OutputType>
     <RootNamespace>FSharp.Control.Reactive</RootNamespace>
     <AssemblyName>FSharp.Control.Reactive</AssemblyName>
-    <TargetFrameworkVersion>v4.0</TargetFrameworkVersion>
+    <TargetFrameworkVersion>v4.5.2</TargetFrameworkVersion>
```
Comment (cloudRoutine) [inline src/FSharp.Control.Reactive.fsproj]: @panesofglass  I'm cool with upgrading the framework version to 4.5.2 and FSharp.Core to 4.0 if you are

**Comment by smoothdeveloper** [comment]
FSharp Core 4.0 works with Framework 4.0.

I'm not using this library at this moment, but I do maintain software that needs to ship for .NET 4.0.

If you are changing the framework just because of most recent FSharp.Core that might not be the right decision.

Maybe tests can be put in a separate project which targets 4.5.2?

**Comment by anton-pt** [comment]
@smoothdeveloper valid point: there should be no need for the main project to change, only the test project. I think it was FsCheck that required 4.5. I will roll back the main project to 4.0 when I get a chance.

**Comment by cloudRoutine** [comment]
@anton-pt once you fix the framework settings this is good to merge

**Comment by anton-pt** [comment]
I made the change, however I ran into a problem. I believe that it's the same issue discussed [here](https://github.com/Reactive-Extensions/Rx.NET/issues/99). It's a bit over my head to fix but, in that discussion, @forki mentions that there is a way to do so?

**Pull request merged**